File: src/nls/nlss.f90
   PROGRAM GALAHAD_NLS_EXAMPLE !  GALAHAD 3.3 - 05/05/2021 AT 14:15 GMT
   USE GALAHAD_NLS_double                     ! double precision version
   IMPLICIT NONE
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )    ! set precision
   TYPE ( NLPT_problem_type ):: nlp
   TYPE ( NLS_control_type ) :: control
   TYPE ( NLS_inform_type ) :: inform
   TYPE ( NLS_data_type ) :: data
   TYPE ( GALAHAD_userdata_type ) :: userdata
   EXTERNAL :: EVALC, EVALJ, EVALH, EVALP
   INTEGER :: s
   INTEGER, PARAMETER :: m = 2, n = 3, j_ne = 4, h_ne = 3, p_ne = 3
   REAL ( KIND = wp ), PARAMETER :: p = 4.0_wp  ! parameter p
! start problem data
   nlp%n = n ; nlp%m = m ; nlp%J%ne = j_ne ; nlp%H%ne = h_ne   ! dimensions
   ALLOCATE( nlp%X( n ), nlp%C( m ) )
   nlp%X = (/ 1.0_wp, 1.0_wp, 1.0_wp /)         ! start from (-1,1,1)
!  sparse co-ordinate storage format
   CALL SMT_put( nlp%J%type, 'COORDINATE', s )  ! Specify co-ordinate storage
   ALLOCATE( nlp%J%val( j_ne ), nlp%J%row( j_ne ), nlp%J%col( j_ne ) )
   nlp%J%row = (/ 1, 2, 1, 2 /)                 ! Jacobian J(x)
   nlp%J%col = (/ 1, 2, 3, 3 /)
   CALL SMT_put( nlp%H%type, 'COORDINATE', s )  ! Specify co-ordinate storage
   ALLOCATE( nlp%H%val( h_ne ), nlp%H%row( h_ne ), nlp%H%col( h_ne ) )
   nlp%H%row = (/ 1, 3, 2 /)                    ! Hessian H(x,y)
   nlp%H%col = (/ 1, 1, 2 /)                    ! NB lower triangle
   ALLOCATE( nlp%P%ptr( m + 1 ), nlp%P%row( p_ne ), nlp%P%val( p_ne ) )
   nlp%P%ptr = (/ 1, 3, 4 /)                    ! start of each column of P
   nlp%P%row = (/ 1, 3, 2 /)                    ! row indices of columns
   ALLOCATE( userdata%real( 1 ) )               ! Allocate space for parameter
   userdata%real( 1 ) = p                       ! Record parameter, p
! problem data complete ; solve using a Gauss-Newton model
   CALL NLS_initialize( data, control, inform ) ! Initialize control params
   control%subproblem_direct = .TRUE.           ! directly solve model problem
   control%model = 3                            ! Gauss-Newton model
   control%jacobian_available = 2               ! Jacobian is available
   inform%status = 1                            ! set for initial entry
   CALL NLS_solve( nlp, control, inform, data, userdata, eval_C = EVALC,       &
              eval_J = EVALJ, eval_H = EVALH )  ! Solve problem
   IF ( inform%status == 0 ) THEN               ! Successful return
     WRITE( 6, "( ' NLS: ', I0, ' iterations -',                               &
    &     ' optimal objective value =',                                        &
    &       ES12.4, /, ' Optimal solution = ', ( 5ES12.4 ) )" )                &
     inform%iter, inform%obj, nlp%X
   ELSE                                         ! Error returns
     WRITE( 6, "( ' NLS_solve exit status = ', I6 ) " ) inform%status
   END IF
! now solve using a Newton model
   control%subproblem_direct = .TRUE.           ! directly solve model problem
   control%model = 4                            ! Change to Newton model
   control%hessian_available = 2                ! Hessian is available
   nlp%X = (/ 1.0_wp, 1.0_wp, 1.0_wp /)         ! start from (-1,1,1)
   inform%status = 1                            ! set for initial entry
   CALL NLS_solve( nlp, control, inform, data, userdata, eval_C = EVALC,       &
              eval_J = EVALJ, eval_H = EVALH )  ! Solve problem
   IF ( inform%status == 0 ) THEN               ! Successful return
     WRITE( 6, "( ' NLS: ', I0, ' iterations -',                               &
    &     ' optimal objective value =',                                        &
    &       ES12.4, /, ' Optimal solution = ', ( 5ES12.4 ) )" )                &
     inform%iter, inform%obj, nlp%X
   ELSE                                         ! Error returns
     WRITE( 6, "( ' NLS_solve exit status = ', I6 ) " ) inform%status
   END IF
! finally solve using a tensor Gauss Newton model
   control%subproblem_direct = .TRUE.           ! directly solve model problem
   control%model = 6                            ! Change to tensor-GN model
   nlp%X = (/ 1.0_wp, 1.0_wp, 1.0_wp /)         ! start from (-1,1,1)
   inform%status = 1                            ! set for initial entry
   CALL NLS_solve( nlp, control, inform, data, userdata, eval_C = EVALC,       &
                   eval_J = EVALJ, eval_H = EVALH,                             &
                   eval_HPRODS = EVALP )        ! Solve problem
   IF ( inform%status == 0 ) THEN               ! Successful return
     WRITE( 6, "( ' NLS: ', I0, ' iterations -',                               &
    &     ' optimal objective value =',                                        &
    &       ES12.4, /, ' Optimal solution = ', ( 5ES12.4 ) )" )                &
     inform%iter, inform%obj, nlp%X
   ELSE                                         ! Error returns
     WRITE( 6, "( ' NLS_solve exit status = ', I6 ) " ) inform%status
   END IF
   CALL NLS_terminate( data, control, inform )  ! delete internal workspace
   DEALLOCATE( nlp%X, nlp%G, nlp%J%val, nlp%J%row, nlp%J%col, userdata%real )
   DEALLOCATE( nlp%H%val, nlp%H%row, nlp%H%col )
   DEALLOCATE( nlp%P%val, nlp%P%row, nlp%P%ptr )
   END PROGRAM GALAHAD_NLS_EXAMPLE

   SUBROUTINE EVALC( status, X, userdata, C )   ! residual
   USE GALAHAD_USERDATA_double
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   INTEGER, INTENT( OUT ) :: status
   REAL ( KIND = wp ), DIMENSION( : ),INTENT( IN ) :: X
   REAL ( KIND = wp ), DIMENSION( : ),INTENT( OUT ) :: C
   TYPE ( GALAHAD_userdata_type ), INTENT( INOUT ) :: userdata
   REAL ( KIND = wp ) :: p
   p = userdata%real( 1 )
   C( 1 ) = X( 3 ) * X( 1 ) ** 2 + P
   C( 2 ) = X( 2 ) ** 2 + X( 3 )
   status = 0
   RETURN
   END SUBROUTINE EVALC

   SUBROUTINE EVALJ( status, X, userdata, J_val )    ! Jacobian
   USE GALAHAD_USERDATA_double
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   INTEGER, INTENT( OUT ) :: status
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( IN ) :: X
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( OUT ) :: J_val
   TYPE ( GALAHAD_userdata_type ), INTENT( INOUT ) :: userdata
   REAL ( KIND = wp ) :: p
   p = userdata%real( 1 )
   J_val( 1 ) = 2.0_wp * X( 1 ) * X( 3 )
   J_val( 2 ) = 2.0_wp * X( 2 )
   J_val( 3 ) = X( 1 ) ** 2
   J_val( 4 ) = 1.0_wp
   status = 0
   RETURN
   END SUBROUTINE EVALJ

   SUBROUTINE EVALH( status, X, Y, userdata, H_val ) ! scaled Hessian
   USE GALAHAD_USERDATA_double
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   INTEGER, INTENT( OUT ) :: status
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( IN ) :: X, Y
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( OUT ) :: H_val
   TYPE ( GALAHAD_userdata_type ), INTENT( INOUT ) :: userdata
   REAL ( KIND = wp ) :: p
   p = userdata%real( 1 )
   H_val( 1 ) = 2.0_wp * X( 3 ) * Y( 1 )
   H_val( 2 ) = 2.0_wp * X( 1 ) * Y( 1 )
   H_val( 3 ) = 2.0_wp * Y( 2 )
   status = 0
   RETURN
   END SUBROUTINE EVALH

   SUBROUTINE EVALP( status, X, V, userdata, P_val, got_h )
   USE GALAHAD_USERDATA_double
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   INTEGER, INTENT( OUT ) :: status
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( IN ) :: X
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( INOUT ) :: P_val
   REAL ( KIND = wp ), DIMENSION( : ), INTENT( IN ) :: V
   TYPE ( GALAHAD_userdata_type ), INTENT( INOUT ) :: userdata
   LOGICAL, OPTIONAL, INTENT( IN ) :: got_h
   P_val( 1 ) = 2.0_wp * ( X( 3 ) * V( 1 ) + X( 1 ) * V( 3 ) )
   P_val( 2 ) = 2.0_wp * X( 1 ) * V( 1 )
   P_val( 3 ) = 2.0_wp * V( 2 )
   status = 0
   RETURN
   END SUBROUTINE EVALP

File: src/nls/C/nlss.c
/* nlss.c */
/* Examples for the NLS C interface using C sparse matrix indexing */
/* Basic tests of models used, direct access */
/* Jari Fowkes & Nick Gould, STFC-Rutherford Appleton Laboratory, 2021 */

#include <stdio.h>
#include <math.h>
#include "galahad_precision.h"
#include "galahad_nls.h"

#define max(a,b)  \
({  __typeof__ (a) _a = (a); \
    __typeof__ (b) _b = (b); \
    _a > _b ? _a : _b; })

// Custom userdata struct
struct userdata_type {
   real_wp_ p;
};

// Function prototypes

int res( int n, int m, const real_wp_ x[], real_wp_ c[], const void * );
int jac( int n, int m, int jne, const real_wp_ x[], real_wp_ jval[], 
         const void * );
int hess( int n, int m, int hne, const real_wp_ x[], const real_wp_ y[], 
          real_wp_ hval[], const void * );
int jacprod( int n, int m, const real_wp_ x[], const bool transpose,  
             real_wp_ u[], const real_wp_ v[], bool got_j, const void * );
int hessprod( int n, int m, const real_wp_ x[], const real_wp_ y[], 
              real_wp_ u[], const real_wp_ v[], bool got_h, const void * );
int rhessprods( int n, int m, int pne, const real_wp_ x[], const real_wp_ v[],
                real_wp_ pval[], bool got_h, const void * );
int scale( int n, int m, const real_wp_ x[], real_wp_ u[], 
           const real_wp_ v[], const void * );
int jac_dense( int n, int m, int jne, const real_wp_ x[], real_wp_ jval[], 
               const void * );
int hess_dense( int n, int m, int hne, const real_wp_ x[], const real_wp_ y[],
                real_wp_ hval[], const void * );
int rhessprods_dense( int n, int m, int pne, const real_wp_ x[], 
                      const real_wp_ v[], real_wp_ pval[], bool got_h,
                      const void * );

int main(void) {

    // Derived types
    void *data;
    struct nls_control_type control;
    struct nls_inform_type inform;

    // Set user data
    struct userdata_type userdata;
    userdata.p = 1.0;

    // Set problem data
    int n = 2; // # variables
    int m = 3; // # residuals
    int j_ne = 5; // Jacobian elements
    int h_ne = 2; // Hesssian elements
    int p_ne = 2; // residual-Hessians-vector products elements
    int J_row[] = {0, 1, 1, 2, 2}; // Jacobian J
    int J_col[] = {0, 0, 1, 0, 1}; // 
    int J_ptr[] = {0, 1, 3, 5};    // row pointers
    int H_row[] = {0, 1};          // Hessian H
    int H_col[] = {0, 1};          // NB lower triangle
    int H_ptr[] = {0, 1, 2};       // row pointers
    int P_row[] = {0, 1};          // residual-Hessians-vector product matrix
    int P_ptr[] = {0, 1, 2, 2};    // column pointers

    // Set storage
    real_wp_ x[n]; // variables
    real_wp_ g[n]; // gradient
    real_wp_ c[m]; // residual
    real_wp_ y[m]; // multipliers
    real_wp_ W[] = {1.0, 1.0, 1.0}; // weights
    char st;
    int status;

    printf(" C sparse matrix indexing\n");

    printf("basic tests of models used, direct access\n\n");

    // ============== Gauss-Newton model ==================

    // Initialize NLS
    nls_initialize( &data, &control, &inform );

    // Set user-defined control options
    control.f_indexing = false; // C sparse matrix indexing
    //control.print_level = 1;
    control.jacobian_available = 2; 
    control.hessian_available = 2;
    control.model = 3;
    x[0] = x[1] = x[2] = 1.5; // starting point

    nls_import( &control, &data, &status, n, m, 
                "sparse_by_rows", j_ne, NULL, J_col, J_ptr,
                "absent", h_ne, NULL, NULL, NULL,
                "absent", p_ne, NULL, NULL, NULL, W ); 
    nls_solve_with_mat( &data, &userdata, &status,
                        n, m, x, c, g, res, j_ne, jac, 
                        h_ne, NULL, p_ne, NULL );

    nls_information( &data, &inform, &status );

    if(inform.status == 0){
        printf(" %i Gauss-Newton iterations. Optimal objective value = %5.2f"
               " status = %1i\n", 
               inform.iter, inform.obj, inform.status);
    }else{
        printf(" NLS_solve exit status = %1i\n", inform.status);
    }
    // Delete internal workspace
    nls_terminate( &data, &control, &inform );

    // ================= Newton model =====================

    // Initialize NLS
    nls_initialize( &data, &control, &inform );

    // Set user-defined control options
    control.f_indexing = false; // C sparse matrix indexing
    //control.print_level = 1;
    control.jacobian_available = 2; 
    control.hessian_available = 2;
    control.model = 4;
    x[0] = x[1] = x[2] = 1.5; // starting point

    nls_import( &control, &data, &status, n, m, 
                "sparse_by_rows", j_ne, NULL, J_col, J_ptr,
                "sparse_by_rows", h_ne, NULL, H_col, H_ptr,
                "absent", p_ne, NULL, NULL, NULL, W ); 
    nls_solve_with_mat( &data, &userdata, &status,
                        n, m, x, c, g, res, j_ne, jac, 
                        h_ne, hess, p_ne, NULL );

    nls_information( &data, &inform, &status );

    if(inform.status == 0){
        printf(" %i Newton iterations. Optimal objective value = %5.2f"
               " status = %1i\n", 
               inform.iter, inform.obj, inform.status);
    }else{
        printf(" NLS_solve exit status = %1i\n", inform.status);
    }
    // Delete internal workspace
    nls_terminate( &data, &control, &inform );

    // ============== tensor-Newton model =================

    // Initialize NLS
    nls_initialize( &data, &control, &inform );

    // Set user-defined control options
    control.f_indexing = false; // C sparse matrix indexing
    //control.print_level = 1;
    control.jacobian_available = 2; 
    control.hessian_available = 2;
    control.model = 6;
    x[0] = x[1] = x[2] = 1.5; // starting point

    nls_import( &control, &data, &status, n, m, 
                "sparse_by_rows", j_ne, NULL, J_col, J_ptr,
                "sparse_by_rows", h_ne, NULL, H_col, H_ptr,
                "sparse_by_columns", p_ne, P_row, NULL, P_ptr, W );
    nls_solve_with_mat( &data, &userdata, &status,
                        n, m, x, c, g, res, j_ne, jac, 
                        h_ne, hess, p_ne, rhessprods );

    nls_information( &data, &inform, &status );

    if(inform.status == 0){
        printf(" %i tensor-Newton iterations. Optimal objective value = %5.2f"
               " status = %1i\n", 
               inform.iter, inform.obj, inform.status);
    }else{
        printf(" NLS_solve exit status = %1i\n", inform.status);
    }
    // Delete internal workspace
    nls_terminate( &data, &control, &inform );
}

// compute the residuals
int res( int n, int m, const real_wp_ x[], real_wp_ c[], const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    real_wp_ p = myuserdata->p;
    c[0] = pow(x[0],2.0) + p;
    c[1] = x[0] + pow(x[1],2.0);
    c[2] = x[0] - x[1];
    return 0;
}

// compute the Jacobian
int jac( int n, int m, int jne, const real_wp_ x[], real_wp_ jval[], 
         const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    jval[0] = 2.0 * x[0];
    jval[1] = 1.0;
    jval[2] = 2.0 * x[1];
    jval[3] = 1.0;
    jval[4] = - 1.0;
    return 0;
}

// compute the Hessian
int hess( int n, int m, int hne, const real_wp_ x[], const real_wp_ y[], 
           real_wp_ hval[], const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    hval[0] = 2.0 * y[0];
    hval[1] = 2.0 * y[1];
    return 0;
}

// compute Jacobian-vector products
int jacprod( int n, int m, const real_wp_ x[], const bool transpose, 
             real_wp_ u[], const real_wp_ v[], bool got_j, 
             const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    if (transpose) {
      u[0] = u[0] + 2.0 * x[0] * v[0] + v[1] + v[2];
      u[1] = u[1] + 2.0 * x[1] * v[1] - v[2];
    }else{
      u[0] = u[0] + 2.0 * x[0] * v[0];
      u[1] = u[1] + v[0]  + 2.0 * x[1] * v[1];
      u[2] = u[2] + v[0] - v[1];
    }
    return 0;
}

// compute Hessian-vector products
int hessprod( int n, int m, const real_wp_ x[], const real_wp_ y[], 
              real_wp_ u[], const real_wp_ v[], bool got_h, 
              const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    u[0] = u[0] + 2.0 * y[0] * v[0];
    u[1] = u[1] + 2.0 * y[1] * v[1];
    return 0;
}

// compute residual-Hessians-vector products
int rhessprods( int n, int m, int pne, const real_wp_ x[], const real_wp_ v[],
                real_wp_ pval[], bool got_h, const void *userdata ){
    struct userdata_type *myuserdata = ( struct userdata_type * ) userdata;
    pval[0] = 2.0 * v[0];
    pval[1] = 2.0 * v[1];
    return 0;
}






















